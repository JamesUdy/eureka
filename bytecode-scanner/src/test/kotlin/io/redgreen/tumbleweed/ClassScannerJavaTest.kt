package io.redgreen.tumbleweed

import io.redgreen.tumbleweed.samples.LambdaAccessingField
import io.redgreen.tumbleweed.samples.MethodReadingField
import io.redgreen.tumbleweed.samples.MethodWritingField
import io.redgreen.tumbleweed.samples.StaticFieldJava
import io.redgreen.tumbleweed.samples.StringConcatenation
import org.approvaltests.Approvals
import org.junit.jupiter.api.Test

class ClassScannerJavaTest {
  private val defaultJavaClassLocation = ClassFileLocation(
    compiledClassesDirectory = "../bytecode-samples/build/classes/java/main",
    fqClassName = "replace-java-class-?"
  )

  @Test
  fun `01 - it can scan a class with methods reading a field`() {
    // given
    val methodReadingField = defaultJavaClassLocation.copy(
      fqClassName = MethodReadingField::class.java.name,
    )

    // when
    val classStructure = ClassScanner.scan(methodReadingField.file)

    // then
    Approvals.verify(classStructure.printable)
  }

  @Test
  fun `02 - it can scan a class with methods writing a field`() {
    // given
    val methodWritingField = defaultJavaClassLocation.copy(
      fqClassName = MethodWritingField::class.java.name,
    )

    // when
    val classStructure = ClassScanner.scan(methodWritingField.file)

    // then
    Approvals.verify(classStructure.printable)
  }

  @Test
  fun `03 - it can scan a class with a lambda function accessing a field`() {
    // given
    val methodReadingAndWritingField = defaultJavaClassLocation.copy(
      fqClassName = LambdaAccessingField::class.java.name,
    )

    // when
    val classStructure = ClassScanner.scan(methodReadingAndWritingField.file)

    // then
    Approvals.verify(classStructure.printable)
  }

  @Test
  fun `04 - it can scan a class with string concatenation`() {
    // given
    val stringConcatenation = defaultJavaClassLocation.copy(
      fqClassName = StringConcatenation::class.java.name,
    )

    // when
    val classStructure = ClassScanner.scan(stringConcatenation.file)

    // then
    Approvals.verify(classStructure.printable)
  }

  @Test
  fun `04 - it can scan a class with static field`() {
    // given
    val staticFieldJava = defaultJavaClassLocation.copy(
      fqClassName = StaticFieldJava::class.java.name,
    )

    // when
    val classStructure = ClassScanner.scan(staticFieldJava.file)

    // then
    Approvals.verify(classStructure.printable)
  }
}
