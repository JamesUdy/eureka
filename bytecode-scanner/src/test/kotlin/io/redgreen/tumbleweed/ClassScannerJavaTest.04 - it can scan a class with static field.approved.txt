Package: io.redgreen.tumbleweed.samples
Class: StaticFieldJava
Extends: java.lang.Object
Fields:
  - java.lang.String CONSTANT
Methods:
  - void <init>()
  - void main([]java.lang.String)
Reads:
  - void main([]java.lang.String) -> java.lang.String CONSTANT
